### PR TITLE
Adds microlabs

### PR DESCRIPTION
**Pull request by John-Candlebury**
### Summary

```SUMMARY: Content "Adds microlabs, small 4x4 labs that open only to the subway network."```

#### Purpose of change

Adds some rewards for subway spelunking, and more midgame looting areas. I also want these for the Hub 01 questline.

#### Describe the solution

Much like the greater labs, microlabs are randomly genned based on a system of tiles. Unlike labs,  the tiles match through their corner rooms, allowing for a higher perceived randomness and more labyrinthine layouts. They are also implemented fully through json, so adding more content is just a matter of adding new mapgen entries and overmap specials.

Because a picture is worth a thousand words, here's a picture of the average microlab:

image

You might notice that it doesn't have a finale room, instead it opts for both a higher loot amount overall, and may have occasional special rooms that have robotics/mutagen/bionic rewards. The one in the picture has a special cloning room (which doesn't do anything but that's beside the point).

It also has for a much higher density of zombie spawns, probable because people got more easily trapped down there. So get some shotguns or something.

#### Additional context

Thanks to ralreegorganon for coding the subway connections, I couldn't have figured them out by myself.

### CODE REVIEW

File: data/json/mapgen/microlab/microlab_generic.json [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+        "  |cci |FFF|66cccc|   = ",
+        "2|||||||||||||||||||||||",
+        "                        ",
+        "                        ",
+        "||22((((((|||||22|||||||",
+        "          |^hd    rrr|  ",
+        "cc  cc  cc|  d      r|  ",
+        "cc  cc  cc|       rrr|  ",
+        "      YY  |W        W|  ",
+        "||   |6tcc||555||555||2|",
```
Comment (anothersimulacrum) [inline data/json/mapgen/microlab/microlab_generic.json]: There's no way to open these doors.

File: data/json/mapgen/microlab/microlab_special_tiles.json [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        "||-0 0 0 0 0 0 0 0 0 -||",
+        "c|-0 0 0 0 0 0 0 0 0 -|F",
+        " |--------------------|F",
+        " |||||||||||||||||||||||",
+        " cccc|c   |^^|    |  cc ",
+        "    c|c   2  2    |     "
+      ],
+      "palettes": [ "microlab" ],
+      "terrain": { "0": "t_vat", "G": "t_card_science" },
+      "monster": { "T": { "monster": "mon_turret" } },
```
Comment (ymber) [inline data/json/mapgen/microlab/microlab_special_tiles.json]:
```suggestion
      "monster": { "T": { "monster": "mon_turret_rifle" } },
```
9mm turrets are being phased out. See #33436.

File: data/json/mapgen/microlab/microlab_generic.json [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+        " th6|FFFFF|  | | l|^^|  ",
+        "|||||||||||  |2|||||||2|",
+        "     |ccc |    |#####|  ",
+        " ccc |c   |    |#####|  ",
+        " c6c |ccc |YY|||||||||c ",
+        " ccc |c   2  |FFcc|F cc ",
+        "     |c   |  |    |     "
+      ],
+      "palettes": [ "microlab" ],
+      "monster": { "T": { "monster": "mon_turret" }, "M": { "monster": "mon_zombie_bio_op" } }
```
Comment (ymber) [inline data/json/mapgen/microlab/microlab_generic.json]:
```suggestion
      "monster": { "T": { "monster": "mon_turret_rifle" }, "M": { "monster": "mon_zombie_bio_op" } }
```

**Comment by tenmillimaster** [comment]
Wow, this definitely looks like the work of a variously at times underfunded and overfunded agency.
Nice!

**Comment by Night-Pryanik** [comment]
I don't know if these labs could be called "micro". "Mini" at best.

Aside from naming issues, I'd like to mention serious current problem with all our labs, micro variant included. Placing all versions of labs (mutations, robotics, aliens, bionics, weaponry etc) in one building is BAD. I think there are no all-in-one labs IRL. I think we need to strictly differentiate all kinds of labs from each other. This will encourage players to explore other labs because they won't find all types of high-level loot in one single lab anymore.

File: data/json/mapgen/microlab/microlab_special_tiles.json (hunk 2 above)
Comment (John-Candlebury) [reply data/json/mapgen/microlab/microlab_special_tiles.json]: True I wrote these specifics maps more than a year ago, and forgot to update

**Comment by Zireael07** [comment]
@Night-Pryanik: You make a very good point, but I believe it's outside this PR's scope.

**Comment by Night-Pryanik** [comment]
Well, this PR in its current state is adding another case of all-in-one labs, and I believe we shouldn't worsen the overall situation.

**Comment by Zireael07** [comment]
I don't think it worsens the overall situation, given that it gives people a real reason to go wandering around the sewers, which were a bit underwhelming.

**Comment by Night-Pryanik** [comment]
It worsens the current "labs" situation, not the "sewers" situation.

**Comment by John-Candlebury** [comment]
The lab added in here is mostly the mundane chemical/biological variant, it barely has high-level, "sci-fi" loot outside the few rooms contained in microlab_special.json. Unlike normal labs they also have no experimental weapons/teleporters/alien tech rooms. So looting one or two of these wont come close to giving you all the neat tools a late game character needs.

As for the special rooms I think that both the mutagen and cloning rooms both fit in the "biological lab"  theme these have going on. The rare robotics room can properly justified as a basic maintenance location, since the labs can have robotic security and the rooms only spawn basic metallurgic tools and electronic components.  

This only leaves the autodoc room as an outlier. I can remove it rather easily, but Im not sure its warranted either, zombie scientists have implants, and theres no strong reason to why these couldnt be installed in site.

---

Dont get me wrong I would love more thematic labs, but as of now we dont have good mapgen tools to do this, at least not without a lot of duplicating mapgen deffinitions, since theres no good way to define item spawns without having to overwrite pretty much everything 

If someone wants to implement a method that allows setting the item_spawns either on overmap_terrain  definition or through ter_furn_transforms, know that I'm willing to churn out a ton of lab content to make specialized labs a real thing.

**Comment by ymber** [comment]
Generally these feel pretty good. Loot spawns are about right and they're an interesting place to explore. The ratio of researchers to security is unduly skewed towards security and I found a mapgen bug while I was exploring it where a table was spawning on the same tile as a wall.

**Comment by John-Candlebury** [comment]
Revised the spawn list, should have a better chance to spawn the lab flavored foes right now.

**Pull request merged**